### PR TITLE
[5.5] Ensure uncaught exceptions are returned in JSON if required

### PR DESCRIPTION
**Pull request by laurencei**
The current framework behavior is to return full html in uncaught exceptions (such as a 500 internal error). 

In `5.5/master` there are already some changes which will forcibly return JSON if `expectsJSON()` is `true` - but only if `app.debug = true`.

I propose that if `expectsJSON()` is `true` - then we should always be returning JSON regardless of the debug setting.

### CODE REVIEW

File: src/Illuminate/Foundation/Exceptions/Handler.php [hunk 1]
```diff
@@ -220,12 +218,18 @@ protected function prepareJsonResponse($request, Exception $e)
...
-            'file' => $e->getFile(),
-            'line' => $e->getLine(),
-            'trace' => $e->getTrace(),
-        ], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), $status, array_merge($headers, [
+        if (config('app.debug')) {
+            $response = [
+                'message' => $e->getMessage(),
+                'file' => $e->getFile(),
+                'line' => $e->getLine(),
+                'trace' => $e->getTrace(),
```
Comment (iflamed) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: Sorry, I don't think we should response a different json in debug mode and the production mode. 

Further more, the exception code (getCode()) may have much more valuable.

May be we should just hide the trace for some security consideration. Also we should consider the translation for the message.

**Review by iflamed — CHANGES_REQUESTED** [review]
See the comment. The security, the exception code, the message translation is important.

File: src/Illuminate/Foundation/Exceptions/Handler.php (hunk 1 above)
Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: > Sorry, I don't think we should response a different json in debug mode and the production mode.

Taylor already added JSON stuff for debug in the master branch. This PR is not adding it.

We are just putting JSON for production when it was requested via AJAX.

**Comment by laurencei** [comment]
I looked around - there dont seem to be any framework tests for the Exceptions Handler?

I've added some tests to start some coverage, at least for these changes.

File: src/Illuminate/Foundation/Exceptions/Handler.php (hunk 1 above)
Comment (iflamed) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: You should just hide the trace in production. Or replace it with the exception's getCode. I think the translation for the message is really important. Could you add these code ?

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: Translation of exceptions should probably be a different PR though - as that is a separate to what is occurring in this PR?

**Comment by m1guelpf** [comment]
@laurencei https://github.com/laravel/framework/pull/18731 ?

File: src/Illuminate/Foundation/Exceptions/Handler.php (hunk 1 above)
Comment (iflamed) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: Maybe, but it is relate these code, about the exception. It doesn't matter how the code work. Please consider the exception code, the translation. The file name and the exception line shown in production may not be safe for the server.

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: > The file name and the exception line shown in production may not be safe for the server.

Correct - we only return the detailed information if `debug` is true.

**Comment by laurencei** [comment]
@m1guelpf - looks like we both created the PRs at similar times.

This PR is a different approach and takes into account the upcoming changes in `5.5/master`.

**Comment by thecrypticace** [comment]
This is something I have to add to _every_ laravel app I write that has an API. It would be nice for something like this to be In core.

File: src/Illuminate/Foundation/Exceptions/Handler.php [hunk 2]
```diff
@@ -220,12 +218,18 @@ protected function prepareJsonResponse($request, Exception $e)
...
+            ];
+        } else {
+            $response['message'] = $this->isHttpException($e) ? $e->getMessage() : 'Server Error';
```
Comment (JosephSilber) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: I believe this to be clearer:

```php
$response = [
    'message' => $this->isHttpException($e) ? $e->getMessage() : 'Server Error',
];
```

...since we're only now creating the response.

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: @JosephSilber - yeah, I was wondering about that. I've pushed an update.

File: src/Illuminate/Foundation/Exceptions/Handler.php [hunk 3]
```diff
@@ -139,9 +139,7 @@ public function render($request, Exception $e)
             return $this->convertValidationExceptionToResponse($e, $request);
         }
 
-        return $request->expectsJson() && config('app.debug')
-                        ? $this->prepareJsonResponse($request, $e)
-                        : $this->prepareResponse($request, $e);
+        return $request->expectsJson() ? $this->prepareJsonResponse($request, $e) : $this->prepareResponse($request, $e);
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: Please leave this on multiple lines how it was.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Foundation/Exceptions/Handler.php (hunk 3 above)
Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: @GrahamCampbell - ok - fixed

File: src/Illuminate/Foundation/Exceptions/Handler.php [hunk 4]
```diff
@@ -220,12 +220,20 @@ protected function prepareJsonResponse($request, Exception $e)
...
+                'line' => $e->getLine(),
+                'trace' => $e->getTrace(),
+            ];
+        } else {
+            $response = [
+                'message' => $this->isHttpException($e) ? $e->getMessage() : 'Server Error',
+            ];
+        }
+
+        return response(json_encode($response, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES), $status, array_merge($headers, [
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: So we actually can only return JSON here, why not just:
```php
return new JsonResponse($response, $status, $headers, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
``` 
`@return` doblock should be changed accordingly

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: @lucasmichot - that line of code is basically what was already in the master branch. But I agree your way seems cleaner - so I've updated the PR.

File: src/Illuminate/Foundation/Exceptions/Handler.php [hunk 5]
```diff
@@ -212,22 +213,28 @@ protected function prepareResponse($request, Exception $e)
      *
      * @param  \Illuminate\Http\Request  $request
      * @param  \Exception $e
-     * @return \Symfony\Component\HttpFoundation\Response
+     * @return \Illuminate\Http\JsonResponse
      */
     protected function prepareJsonResponse($request, Exception $e)
     {
         $status = $this->isHttpException($e) ? $e->getStatusCode() : 500;
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: I believe only one condition could be used here ?

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: @lucasmichot - what do you mean?

Comment (lucasmichot) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: I meant that we can actually make this check only once:
```php
if ($this->isHttpException($e)) {
    $status = $e->getStatusCode();
    $headers = $e->getHeaders();
} else {
    $status = 500;
    $headers = [];
}
```

Comment (laurencei) [inline src/Illuminate/Foundation/Exceptions/Handler.php]: Up to Taylor I guess? Those lines were what he wrote already in the master branch that currently exists. I'll leave it as is for this PR to keep changes to a minimum and we can review after?

**Comment by SteamWind** [comment]
Hello,
What if we want those HTML errors back in debug? Is there a way to setup this?

**Comment by laurencei** [comment]
You could override the exception handler - yes
